packages/@aws-containers/aws-ecs-builder/lib/extensions/extension-interfaces.ts

```diff
@@ -0,0 +1,218 @@
+   * The container of this extension. Most extensions have a container, but not
+   * every extension is required to have a container, some extensions may just
+   * modify the properties of the service, or create external resources
+   * connected to the service
+   */
+  public container?: ecs.ContainerDefinition;
+
+  protected parentService!: Service;
+  protected scope!: cdk.Construct;
+
+  // A list of other extensions which want to mutate the
+  // container definition for this extension.
+  protected containerMutatingHooks: ContainerMutatingHook[] = [];
+
+  constructor(name: string) {
+    this.name = name;
+  }
+
+  /**
+   * A hook that allows the extension to add hooks to other
+   * extensions that are registered
+   */
+  public addHooks() { } // tslint:disable-line
+
+  /**
+   * This hook allows another service extension to register a mutating hook for
+   * changing the primary container of this extension. This is primarily used
+   * for the application extension. For example the Firelens extension wants to
+   * be able to modify the settings of the application container to
+   * route logs through Firelens.
+   * @param hook
+   */
+  public addContainerMutatingHook(hook: ContainerMutatingHook) {
+    this.containerMutatingHooks.push(hook);
+  }
+
+  /**
+   * This is a hook which allows extensions to modify the settings of the
+   * task definition prior to it being created. For example App Mesh
+   * extension needs to configure an Envoy proxy in the task definition,
+   * or the application extension wants to set the overall resource for
+   * the task.
+   * @param props - Properties of the task definition to be created
+   */
+  public mutateTaskDefinitionProps(props: ecs.TaskDefinitionProps) {
+    return {
+      ...props,
+    } as ecs.TaskDefinitionProps;
+  }
+
+  /**
+   * A hook that is called for each extension adhead of time to
+   * let it do any initial setup, such as creating resources in
+   * advance.
+   * @param parent - The parent service which this extension has been added to
+   * @param scope - The scope that this extension should create resources in
+   */
+  public prehook(parent: Service, scope: cdk.Construct) {
+    this.parentService = parent;
+    this.scope = scope;
+  }
+
+  /**
+   * Once the task definition is created, this hook is called for each
+   * extension to give it a chance to add containers to the task definition,
+   * change the task definition's role to add permissions, etc
+   * @param taskDefinition - The created task definition to add containers to
+   */
+  public useTaskDefinition(taskDefinition: ecs.TaskDefinition) {
+    taskDefinition = taskDefinition;
+  }
+
+  /**
+   * Once all containers are added to the task definition this hook is
+   * called for each extension to give it a chance to bake its dependency
+   * graph so that its container starts in the right order based on the
+   * other extensions that were enabled
+   */
+  public bakeContainerDependencies() {
```

Comment:
Shouldn't this be `resolveContainerDependencies`?

Reply:
Also, since this is called after each extension is added to check for existing extensions, are we sure there won't be any conflicting logic based on the order of adding each one?